### PR TITLE
feat: add Apple Silicon and MPS support for macOS

### LINKED ISSUE #172: [Feature]: Add Apple Silicon (MPS) support for macOS ML environments
### Is your feature request related to a problem?
Yes.

EnvForge currently focuses on NVIDIA CUDA and AMD ROCm acceleration paths for ML environment provisioning. However, Apple Silicon users (M1/M2/M3/M4 Macs) rely on Metal Performance Shaders (MPS) instead of CUDA/ROCm for GPU acceleration.

As a result, macOS Apple Silicon users may not receive accurate compatibility resolution, environment generation, or verification support for frameworks like PyTorch and TensorFlow.

Problem

EnvForge currently focuses on NVIDIA CUDA and AMD ROCm acceleration paths for ML environment provisioning. However, Apple Silicon users (M1/M2/M3/M4 Macs) rely on Metal Performance Shaders (MPS) instead of CUDA/ROCm for GPU acceleration.

At the moment, macOS Apple Silicon users may not get accurate compatibility resolution or environment generation for frameworks like PyTorch and TensorFlow.

### Describe the solution you'd like

Add first-class Apple Silicon / MPS support across the platform.

This could include:

> CLI Hardware Detection

- Detect Apple Silicon architecture (arm64)

- Detect macOS platform

- Identify MPS acceleration availability

> Compatibility Engine Updates

- Add compatibility rules for:

- PyTorch with MPS backend

- TensorFlow + tensorflow-metal

- Avoid CUDA-specific logic for Apple Silicon targets

> New Environment Profiles

Example:

- pytorch-mps

- tensorflow-metal

> Template Support

Generate macOS-compatible setup scripts for Apple Silicon users.

Example:

python3 -m venv .venv
source .venv/bin/activate
pip install torch torchvision torchaudio

or

pip install tensorflow tensorflow-metal
Verification Support
Add validation checks such as:
torch.backends.mps.is_available()

instead of CUDA-only verification paths.

Why this would help

Apple Silicon machines are widely used by students, developers, and ML practitioners. Adding MPS support would significantly expand EnvForge’s usability beyond CUDA/ROCm ecosystems and improve cross-platform coverage.

### Alternatives Considered

1. **Treat Apple Silicon users as CPU-only**
This would technically allow environment generation on macOS, but it ignores available GPU acceleration through Apple's MPS backend and gives a degraded ML experience.

2. **Let users manually configure framework installations**
Users could manually install PyTorch with MPS support or TensorFlow with `tensorflow-metal`, but this goes against EnvForge’s goal of deterministic, automated environment provisioning.

3. **Generic macOS support without accelerator awareness**
Basic macOS setup generation is possible, but without explicit Apple Silicon/MPS compatibility logic, generated environments may be incomplete or incorrectly assume CUDA-based workflows.

I believe first-class Apple Silicon support is the better long-term solution since it aligns with EnvForge’s cross-platform environment provisioning goals.

### Additional context

PyTorch already supports MPS acceleration on macOS, and TensorFlow provides Metal acceleration through tensorflow-metal.

### PR DESCRIPTION
**Pull request by abhishekKokadwar**
## Summary

Adds first-class Apple Silicon / Metal Performance Shaders (MPS) support to EnvForge for macOS users.

This expands accelerator support beyond NVIDIA CUDA and AMD ROCm to support Apple Silicon ML workflows.

### CLI

* Added macOS detection (`Darwin`)
* Added Apple Silicon (`arm64`) detection
* Added MPS capability detection
* Added MPS GPU detection tests

### Backend

* Added `MACOS` as supported `OSTarget`
* Added MPS compatibility matrix
* Updated OS compatibility rules
* Updated seed schema validation to allow macOS profiles
* Added Apple-specific environment profiles:

  * `pytorch-mps`
  * `tensorflow-metal`

### Templates

Added macOS-specific setup and verification templates:

* `setup_macos.sh.j2`
* `verify_torch_mps.sh.j2`
* `verify_tf_metal.sh.j2`

Updated template engine for OS-aware template selection.

## Testing

Feature-specific tests:

* CLI MPS tests: passed
* Backend MPS compatibility tests: passed

Full regression:

* Backend: 191/191 passed
* CLI: 81/81 passed

## Notes

* Validation performed on Windows development environment
* Docker/local seed auth mismatch observed during local setup, but seed validation and test coverage passed
* No physical Apple Silicon hardware validation performed

## Summary by CodeRabbit

* **New Features**
  * Added macOS / Apple Silicon support with GPU acceleration for PyTorch MPS and TensorFlow Metal.
  * New environment profiles for pytorch-mps and tensorflow-metal.
  * macOS-specific setup and verification scripts for hardware-acceleration validation.
  * Enhanced OS and GPU detection to recognize macOS systems and report MPS/Metal.

* **Bug Fixes**
  * Diagnose/OS-target logic updated to correctly identify macOS.

* **Tests**
  * Added comprehensive tests for macOS, MPS/Metal support, and detection logic.

[Review Change Stack](https://app.coderabbit.ai/change-stack/rishabh0510rishabh/EnvForage/pull/207?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by rishabh0510rishabh** [comment]
can you Tag issue you have been assigned to for this?

**Comment by abhishekKokadwar** [comment]
> can you Tag issue you have been assigned to for this?

Fixes #172

**Comment by rishabh0510rishabh** [comment]
hey @Abhishek86798 the backend tests have failed please recheck the code
